python/pyarrow/dataset.py

```diff
@@ -592,9 +596,13 @@ def dataset(source, schema=None, format=None, filesystem=None,
         selector_ignore_prefixes=ignore_prefixes
     )
 
+    import io
+
     # TODO(kszucs): support InMemoryDataset for a table input
     if _is_path_like(source):
         return _filesystem_dataset(source, **kwargs)
+    elif isinstance(source, (io.BytesIO, FileSource)):
```

Comment:
```suggestion
    elif isinstance(source, (io.BytesIO, FileSource, pa.NativeFile)):
```

?

Reply:
and also `pa.Buffer`

Reply:
I can certainly add these but it is far from a complete and general solution. For example this would not enable you to pass a list of buffers or file like objects other than bytesIO and nativefile

Reply:
Yep, I understand, but added them now in the commit I pushed to get the tests passing (those only use single buffer objects). But certainly agree this should be cleaned up (later).

Reply:
Well for the time being I've added `dataset.py::_is_native_file_wrappable` which includes an explicit and easily extensible class list. Further refactoring can be left for a follow up